### ISSUE TITLE
Improve XAML by removing noise

### ISSUE THREAD

**Issue by ichthus1604** [AvaloniaUI/Avalonia#11906]
**Is your feature request related to a problem? Please describe.**

I'm always frustrated when trying to explain XAML to a web developer.

**Describe the solution you'd like**

Remove all the unnecessary noise from XAML and make it friendlier to everyone.

**Describe alternatives you've considered**

None at the time.

**Additional context**

So, this is the XAML for an Avalonia Hello World:

```XAML
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="MyApp.Window1"
        Title="Window1">
  Welcome to Avalonia!
</Window>
```

That's a lot of noise to explain to somebody who has never used Avalonia nor WPF.
It would be awesome if it was just:

```xaml
<Window Title="Window1">
  Welcome to Avalonia!
</Window>
```

Since none of the other stuff is really required.

 - `xmlns` and `xmlns:x` should really be implicit for every XAML file every, since they're pretty much required.
 - `xmlns:d` and `xmlns:mc` are designer-specific concerns that don't really belong into production source code. Make the designer / previewer handle this implicitly.
 - `x:Class` has 2 pieces of information:
    - the namespace could be implicitly derived by the location of the XAML file relative to the root of the project (following C# namespacing convention). It could still be explicitly added manually to override the convention if required.
    - the class name could be implicitly derived from the XAML file name.

Also: it would be awesome if custom and user controls in an assembly where implicitly mapped to the root namespace instead of having to define a custom XAML namespace for them.

**Comment by maxkatz6** [comment]
I have mixed feelings here.

I kinda agree that default xaml files can be improved. And I personally believe that a huge part of it - design-specific namespaces, can be eliminated altogether. But if we go in details...

> xmlns and xmlns:x should really be implicit for every XAML file ever, since they're pretty much required.

XAML should still be a valid XML. Using nodes with namespaces (x:String...) without actually defining a namespace is not a valid XML. It won't be parsable by typical XML parser as well.
`xmlns="https://github.com/avaloniaui"` specifically could be implicit. The problem is - Avalonia is not the only framework that uses XAML files, and xmlns is typically a way to differentiate that. Saying this, Avalonia switched to AXAML files by default, which is the same file format, but a different extension to avoid Visual Studio WPF bugs. So, possibly only for AXAML it can be implicit.

Now...`xmlns:x` is not really necessary in the default templates. And it ideally can be added by the IDE when any type is referenced from there.

> xmlns:d and xmlns:mc are designer-specific concerns that don't really belong into production source code. The designer / previewer could handle this implicitly.

Agreed. But note, ` d:DesignWidth="800" d:DesignHeight="450"` still needs to be specified. We can assume default values for the previewer and not force them, but there are lots of cases when developers actually want to test their controls/windows without any height/width set. How would they remove previewer size assumptions? 
But returning to my point about "eliminated design namespaces altogether" - it can be Design.Width="800" and Design.Height="450" properties that do not require any special namespace.

> x:Class has 2 pieces of information:

I think this one can be safely assumed by the compiler when both XAML and C# files are of the same type. The question is if it can be a problem with existing use cases. I can't imagine one.

**Comment by maxkatz6** [comment]
Personally, I believe that XAML can be improved, and it's still a better option than inventing a brand-new markup language.

**Comment by rabbitism** [comment]
> Personally, I believe that XAML can be improved, and it's still a better option than inventing a brand-new markup language.

Maybe we can introduce a feature like global using in axaml.

**Comment by maxkatz6** [comment]
@rabbitism it will be even less valid XML this way, though.

**Comment by ichthus1604** [comment]
@maxkatz6 I think explicit XAML namespacing (at least for platform types) is fundamentally wrong. I understand the reasoning and design considerations behind it, but honestly at this point it looks to me as an unfortunate leftover from WPF compatibility.

Why is it `{x:Type}` and not just `{Type}`? Again, the only rationale for this seems to be WPF / other Microsoft stuff compat, which I as an Avalonia user don't value at all to be honest.

I'd rather see the platform advance into better usability and compete in the WASM space rather than staying forever tethered to legacy Microsoft stuff.

Also, even if WPF compat is a must, you could still support both idioms: keep `{x:Type}` for compatibility and enable `{Type}` as a way forward. This could even be toggled by an MSBuild option if needed.

None of my Avalonia-based projects have anything to do with WPF. On the contrary, I'm writing greenfield software (primarily targeting WASM) using Avalonia right now.

If you remove the unnecessary differentiation of namespaces for platform types, then you no longer need `xmlns:x`, and if you implicitly define `xmlns="https://github.com/avaloniaui"`, then you have already removed 90% of the noise in XAML for good, and XAML would *still* be valid XML because you're not using any unimported namespaces.

The only thing left for complete XAML happiness is to consider all assembly-local `AvaloniaObject`s as part of the root namespace, so I don't have to `xmlns:controls="using:MyApp.Controls"` and then litter all my XAML with `controls:` everywhere. Not to mention that this escalates quite badly when you work with ViewModels and Views, which tipically belong to different namespaces, then you end up with

```xaml
<Something
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:viewmodels="using:MyApp.ViewModels"
    xmlns:someotherviewmodels="using:MyApp.SomeOtherViewModels"
    xmlns:views="using:MyApp.Views"
    xmlns:someotherviews="using:MyApp.SomeOtherViews">
    <DataTemplate DataType="{x:Type viewmodels:ViewModel1}">
        <views:View1/>
    </DataTemplate>
    <DataTemplate DataType="{x:Type someotherviewmodels:ViewModel2}">
        <someotherviews:View2/>
    </DataTemplate>
</Something>
```

I think it is pretty obvious that this example above has very unfavorable noise to signal ratio. It could just be:

```xaml
<Something>
    <DataTemplate DataType="{Type ViewModel1}">
        <View1/>
    </DataTemplate>
    <DataTemplate DataType="{Type ViewModel2}">
        <View2/>
    </DataTemplate>
</Something>
```

**Comment by timunie** [comment]
fyi: it's possible to define your own namespace and move all `Views`, `ViewModels`, `Converter` etc into that ns. Can also be the Avalonia one, then you can leave the prefix. See: https://docs.avaloniaui.net/guides/basics/introduction-to-xaml#create-a-common-namespace

**Comment by byme8** [comment]
> it will be even less valid XML this way,
What is the point of keeping it compatible with XML? No one in existence will use an XML parser to parse XAML.

**Comment by Flithor** [comment]
> > it will be even less valid XML this way,
> 
> What is the point of keeping it compatible with XML? No one in existence will use an XML parser to parse XAML.

Dude, XAML parser is parsing XAML as XML!

**Comment by fforjan** [comment]
> To summarize, in my imagination we should be able safely go down to:
> 
> ```
> <Window xmlns="https://github.com/avaloniaui"
>         Design.Width="800" Design.Height="450"
>         Title="Window1">
>   Welcome to Avalonia!
> </Window>
> ```
> 
> Which already is much better.

Why do we need set a design size ? could we have an implicit size by default and only set the value if you need it ?
And technically, I would have prefer to set something like 'lanscape' vs 'portait mode' or couple of other options instead of 'hard coded size' which always seems wrong anyway ?

**Comment by byme8** [comment]
> Dude, XAML parser is parsing XAML by XML schema!

To some degree, you are right. However, after the parsing, it does lots of magic on top of it. Without this magic, XAML is not XAML. So, effectively it is a completely new parse that is based on XML, but not limited by it.

**Comment by byme8** [comment]
> > To summarize, in my imagination we should be able safely go down to:
> > ```
> > <Window xmlns="https://github.com/avaloniaui"
> >         Design.Width="800" Design.Height="450"
> >         Title="Window1">
> >   Welcome to Avalonia!
> > </Window>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Which already is much better.
> 
> Why do we need set a design size ? could we have an implicit size by default and only set the value if you need it ? And technically, I would have prefer to set something like 'lanscape' vs 'portait mode' or couple of other options instead of 'hard coded size' which always seems wrong anyway ?

And potentially it can be done in the preview itself. No need to define such values in XAML.

**Comment by Symbai** [comment]
Changing it would make it even more difficult for developers coming from WPF or MAUI or any other framework using XAML. And one of the reasons why Avalonia has been so successful is because for a typical C# developer, everything is familiar. Nothing is more frustrating then to read tons of documentation pages to get some basics done.

Instead of tampering with namespaces I'd actually enjoy if we have a better intellisense support. Because sometimes its missing and when its there, its not showing all the properties you can access. Its so awful comparing to WPF that I stopped using it and always look at / copy existing code when working on new pages. This would improve working with XAML a lot. Removing namespaces? Well I would put that into "nice to have" but that doesn't make MY life easier and also the life for users who start with Avalonia and then switch to WPF/MAUI or the other way around. 

Another reason against the global mapping of namespaces is that for one page you use something like `xmlns:vm="clr-namespace:YXZ.ViewModels"` and for the other page you use `xmlns:vm="clr-namespace:Abc123.ViewModels"`. This is the same situation why global usings on C# leads to cause conflicts if a name of a class exists both.

Improvements should make developing more easier and not just "looks better because less code". Sometimes more code is better. I can just think of the top level statement C# had introduced just to quickly provide a way to convert it back to old code + even added a checkbox when creating new C# projects if you want to use that or not. Because so many people hated it. Similar idea, similar solution, similar output.

**Comment by ichthus1604** [comment]
> Changing it would make it even more difficult for developers coming from WPF or MAUI or any other framework using XAML

I strongly disagree.

There is nothing "more difficult" in this:

```xaml
<Window Title="Window1">
  Welcome to Avalonia!
</Window>
```

compared to this:

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="MyApp.Window1"
        Title="Window1">
  Welcome to Avalonia!
</Window>
```

Quite the contrary, all that useless noise makes me really want to quit XAML and learn HTML and React just to be able to write markup similar to the first example.

> Nothing is more frustrating then to read tons of documentation pages to get some basics done

Again, there is absolutely no need to read any documentation to understand my first example. Quite the contrary, people need to read tons of documentation just to be able to put up a simple Hello World app with all that amount of useless noise and unnecessary complexity in the markup.

> users who start with Avalonia and then switch to WPF/MAUI

Sorry, WPF is legacy and MAUI is honestly unusable.

Also, as mentioned above, my proposal includes an MSBuildFlag (let's call it `<UseLegacyXaml>True</UseLegacyXaml>` if you want / need to use legacy XAML.

> cause conflicts if a name of a class exists both

This does not happen in practice, and if it does the compiler should fail to build that specific case, which is 0.00001% compared to the 99.9999% of times you do not have any conflict but still need to explicitly specify everything manually, which again leads to useless noise.

> Improvements should make developing more easier and not just "looks better because less code".

My proposal makes development substantially easier by avoiding unnecessary details such as namespaces that are totally not relevant when writing, nor maintaining code.

> even added a checkbox

Yes. People are free to write legacy code with legacy idioms all they want. I personally want a way forward that doesn't force me to write code like it was 2006.

**Comment by cesarchefinho** [comment]
I agree.

the same "reducing noise" ocourred in html when browsers drop xhtml in favor to html5 that stripoed xml and merged svg and mathml in html

the same needs to be made in axaml

**Comment by Sorien** [comment]
Guys, look here https://github.com/AvaloniaUI/Avalonia/issues/2502 for better proposal

**Comment by cesarchefinho** [comment]
I dont like #2502

It is not better than this proposal, 2502 is flutter or qml or other strange thing but not xaml

this proposal maintain xaml and only simplifies markup, improving developper experience.

**Comment by nnnpa31** [comment]
I don't think it's necessary to keep xaml in xml-based format, so I'm relatively supportive of #2502, but it's important to note that there's still a lot of noise in the original proposal for #2502 that could be improved.

**Comment by robloo** [comment]
My comments:

 1. There could probably be some implicit namespaces... if XML supported it
 2. XAML should follow XML because the whole infrastructure has depended on that for almost 20 years
 3. We really need to be investigating new markup formats such as razor syntax to cover these types of modernizations. No one with existing XAML has complaints about this sort of thing. It just new devs coming on board.

So my vote is to leave things as they are for XAML and put effort into a more modern razor/blazor syntax that can be used side-by-side. https://github.com/AvaloniaUI/Avalonia/issues/2502#issuecomment-631845909

**Comment by nnnpa31** [comment]
> So my vote is to leave things as they are for XAML and put effort into a more modern razor/blazor syntax that can be used side-by-side. https://github.com/AvaloniaUI/Avalonia/issues/2502#issuecomment-631845909

Support this proposal, but it is easy to note that all similar proposals, including #11906 and #2502, eventually became unexplored and died just like #2502. Perhaps the Avalonia developers are working on this proposal, but I haven't seen anything about it from the official team in a long time.

**Comment by timunie** [comment]
Yeah, it's not that high on our todo list I guess. Overall, our API should be flexible enough to have a community lib that does provide another syntax. For example: 

https://github.com/wieslawsoltes/NXUI
https://github.com/fsprojects/Avalonia.FuncUI

**Comment by lloydjatkinson** [comment]
MAUI is doing this as per the .NET Conf 2025 stream, and it looks exactly like the example in the first comment.

```xaml
<Window Title="Window1">
  Welcome to Avalonia!
</Window>
```

**Comment by ichthus1604** [comment]
I'm since off of XAML and Avalonia and I've created my own js-based layout system on top of vue3 components, basically:

    <grid rows="1 auto">
        <textbox align-v="center" align-h="right" row="0"/>
 
         <border background="primary" row="1"/>
    </grid>

best of both worlds: sane layout system, works in the browser, super fast rendering, no wasm, no big initial download.
This (and many other similar issues) taught me that if want things done right you have to do them yourself, no one is coming to save you.

Nice to see I've triggered some discussion though.

**Comment by lloydjatkinson** [comment]
I don’t really see the relevance of that to cleaning up XAML namespaces. Maybe this needs to be a new issue then.

**Comment by Athari** [comment]
MAUI for .NET 10 introduced [Implicit and Global XML namespaces](https://learn.microsoft.com/en-us/dotnet/maui/whats-new/dotnet-10?view=net-maui-10.0#implicit-and-global-xml-namespaces) ([source markdown](https://github.com/dotnet/docs-maui/blob/main/docs/whats-new/dotnet-10.md)):

----

.NET 10 Preview 5 introduces a cleaner **XML-namespace experience** for .NET MAUI that wipes out nearly all of the boiler-plate `xmlns:` lines you used to copy-paste at the top of every XAML file.

| What changed                         | How it works                                                                                                                                                                 |
|--------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Project-wide “global” namespaces** | `http://schemas.microsoft.com/dotnet/maui/global` is a new xmlns that you can use to aggregate multiple xmlns together. By default, it contains the maui xmlns (sourcegenerated), {YourNamespace}, and {YourNamespace}.Pages |
| **Backward compatible**              | Existing explicit `xmlns:` mappings still compile; add them only when you need to disambiguate duplicate type names.                                                        |
| **Implicit default namespace** (opt-in)      | When opting in, the compiler now injects `http://schemas.microsoft.com/dotnet/2021/maui` automatically, so you can drop root `xmlns` and `xmlns:x` lines.                               |

...

```xml
<!-- .NET 10 Preview 5 plus opt-in -->
<ContentPage x:Class="MyApp.MainPage">
    <TagView x:DataType="Tag" />
</ContentPage>
```

...

#### Sample `GlobalXmlns.cs`

```csharp
[assembly: XmlnsDefinition("http://schemas.microsoft.com/dotnet/maui/global", "MyApp.Views")]
[assembly: XmlnsDefinition("http://schemas.microsoft.com/dotnet/maui/global", "MyApp.Controls")]
[assembly: XmlnsDefinition("http://schemas.microsoft.com/dotnet/maui/global", "MyApp.Converters")]
```

If you prefer to continue using the xmlns prefixes in your XAML, you can provide default prefixes for them in the `GlobalXmlns.cs` as well.

```csharp
using XmlnsPrefixAttribute = Microsoft.Maui.Controls.XmlnsPrefixAttribute;

[assembly: XmlnsPrefix("MyApp.Controls","controls")]
```

> ✨ **Tip:** You can register third-party libraries here too!

----

Basically, they seem to just jam everything into one global XML namespace using `XmlnsDefinitionAttribute`. But if it works, it works.

**Comment by fitdev** [comment]
Hopefully this will be addressed soon. Really miss such a feature, which is a huge thing for larger projects and prefix standardization.
